# Capstone Project for Coursera
This notebook will be mainly used for capstone project of Coursera online certificate for **IBM Data Science Professional Ceritificate**.

In [3]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
print("Hello Capstone Project Course!")

Hello Capstone Project Course!


### Import Libraries

First import the libraries which will be used in this capstone project.

In [4]:
import json

# !conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim

import requests
from pandas.io.json import json_normalize

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

# !conda install -c conda-forge folium=0.5.0 --yes
import folium
print('Libraries imported.')

Libraries imported.


### Import Data

Next, using **BeautifulSoup** to transform the data in the table on the Wikipedia page into pandas dataframe. At the same time, cleanning dataframe as the requirement stated.

In [5]:
from bs4 import BeautifulSoup

with open('List of postal codes of Canada_ M - Wikipedia.html') as html_file:
    soup = BeautifulSoup(html_file, 'lxml')


In [6]:
table = soup.find('table', class_='wikitable sortable')
table_rows = table.find_all("tr")
table_columns = table.find_all("th")
l = []
for tr in table_rows:
    td = tr.find_all("td")
    row = [tr.text for tr in td]
    if row and row[1] != 'Not assigned':
        if row[2] == 'Not assigned\n':
            row[2] = row[1]
        l.append(row)
cols = [c.text.replace('\n', '') for c in table_columns]
df = pd.DataFrame(l, columns=cols).replace('\n', '', regex=True)
df = df.groupby(['Borough', 'Postcode'], sort=False).agg(lambda x: ', '.join(set(x.dropna()))).reset_index()
df = df[cols].replace('\n', '', regex=True)
df.head()

Postcode           Borough                     Neighbourhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto         Harbourfront, Regent Park
3      M6A        North York  Lawrence Heights, Lawrence Manor
4      M7A      Queen's Park                      Queen's Park

In [7]:
df.shape

(103, 3)

As the result of **shape** method, the size of this transformed table is **(103, 3)**.

Next step is add the **Latitude** and **Longitude** to the dataframe.

In [8]:
ll = pd.read_csv('Geospatial_Coordinates.csv')
ll.columns = ['Postcode', 'Latitude', 'Longitude']
cols = ['Postcode']
df_withll = df.join(ll.set_index(cols), on=cols)

### Use geopy library to get the latitude and longitude values of Toronto

In [9]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


### Create a map of Ontario with neighborhoods superimposed on top

Create map of Toronto using latitude and longitude values, and then add markers to map

In [10]:

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_withll['Latitude'], df_withll['Longitude'], df_withll['Borough'], df_withll['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Then filtering the **Borough** by contains **Toronto**

In [11]:
df_tor = df_withll[df_withll['Borough'].str.contains("Toronto")]
df_tor.head()

Postcode           Borough              Neighbourhood   Latitude  Longitude
2       M5A  Downtown Toronto  Harbourfront, Regent Park  43.654260 -79.360636
9       M5B  Downtown Toronto   Garden District, Ryerson  43.657162 -79.378937
15      M5C  Downtown Toronto             St. James Town  43.651494 -79.375418
19      M4E      East Toronto                The Beaches  43.676357 -79.293031
20      M5E  Downtown Toronto                Berczy Park  43.644771 -79.373306

Creating map of Toronto using only Borough that contains Toronto.

In [12]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df_tor['Latitude'], df_tor['Longitude'], df_tor['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

### Define Foursquare Credentials and Version

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

In [13]:
CLIENT_ID = 'M5NXRZI5KMSVEMTB1FHXAKUUXF5ODQFPRDMJQEZ053TRIV4W' # your Foursquare ID
CLIENT_SECRET = 'HBRGKYXZTPLCSDZPS4HQEEGZFOTJ2LKL4N5XMURF0OW3URRB' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: M5NXRZI5KMSVEMTB1FHXAKUUXF5ODQFPRDMJQEZ053TRIV4W
CLIENT_SECRET:HBRGKYXZTPLCSDZPS4HQEEGZFOTJ2LKL4N5XMURF0OW3URRB


In [14]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude, 
    longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=M5NXRZI5KMSVEMTB1FHXAKUUXF5ODQFPRDMJQEZ053TRIV4W&client_secret=HBRGKYXZTPLCSDZPS4HQEEGZFOTJ2LKL4N5XMURF0OW3URRB&v=20180605&ll=43.653963,-79.387207&radius=500&limit=100'

In [15]:
results = requests.get(url).json()

In [16]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [17]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(5)

name        categories        lat        lng
0    Downtown Toronto      Neighborhood  43.653232 -79.385296
1             Japango  Sushi Restaurant  43.655268 -79.385165
2         Cafe Plenty              Café  43.654571 -79.389450
3           Poke Guys        Poke Place  43.654895 -79.385052
4  Sansotei Ramen 三草亭  Ramen Restaurant  43.655157 -79.386501

In [18]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

71 venues were returned by Foursquare.


Then explore neighborhoods in Toronto

In [19]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [20]:
toronto_venues = getNearbyVenues(names=df_tor['Neighbourhood'],
                                   latitudes=df_tor['Latitude'],
                                   longitudes=df_tor['Longitude']
                                  )



Harbourfront, Regent Park
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Toronto Islands, Union Station
Trinity, Little Portugal
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North, Forest Hill West
The Junction South, High Park
North Toronto West
North Midtown, Yorkville, The Annex
Parkdale, Roncesvalles
Davisville
Harbord, University of Toronto
Runnymede, Swansea
Summerhill East, Moore Park
Kensington Market, Grange Park, Chinatown
Deer Park, Summerhill West, South Hill, Forest Hill SE, Rathnelly
Bathurst Quay, King and Spadina, Island airport, Harbourfront West, South Niagara, Railway Lands, CN Tower
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown, St. James Town
Fir

#### See the size of toronto_venues dataframe

In [21]:
print(toronto_venues.shape)
toronto_venues.head()

(1689, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Harbourfront, Regent Park               43.65426              -79.360636   
1  Harbourfront, Regent Park               43.65426              -79.360636   
2  Harbourfront, Regent Park               43.65426              -79.360636   
3  Harbourfront, Regent Park               43.65426              -79.360636   
4  Harbourfront, Regent Park               43.65426              -79.360636   

                                             Venue  Venue Latitude  \
0                                 Roselle Desserts       43.653447   
1                                    Tandem Coffee       43.653559   
2  Toronto Cooper Koo Family Cherry St YMCA Centre       43.653191   
3                               Morning Glory Cafe       43.653947   
4                                   Impact Kitchen       43.656369   

   Venue Longitude        Venue Category  
0       -79.362017                Bakery  
1       -79.361809           Coffee Shop  
2       -79.357947  Gym / Fitness Center  
3       -79.361149        Breakfast Spot  
4       -79.356980            Restaurant

In [22]:
toronto_venues.groupby('Neighborhood').count().head()

Neighborhood Latitude  \
Neighborhood                                                                
Bathurst Quay, King and Spadina, Island airport...                     16   
Berczy Park                                                            56   
Brockton, Parkdale Village, Exhibition Place                           21   
Business Reply Mail Processing Centre 969 Eastern                      18   
Cabbagetown, St. James Town                                            44   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Bathurst Quay, King and Spadina, Island airport...                      16   
Berczy Park                                                             56   
Brockton, Parkdale Village, Exhibition Place                            21   
Business Reply Mail Processing Centre 969 Eastern                       18   
Cabbagetown, St. James Town                                             44   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Bathurst Quay, King and Spadina, Island airport...     16              16   
Berczy Park                                            56              56   
Brockton, Parkdale Village, Exhibition Place           21              21   
Business Reply Mail Processing Centre 969 Eastern      18              18   
Cabbagetown, St. James Town                            44              44   

                                                    Venue Longitude  \
Neighborhood                                                          
Bathurst Quay, King and Spadina, Island airport...               16   
Berczy Park                                                      56   
Brockton, Parkdale Village, Exhibition Place                     21   
Business Reply Mail Processing Centre 969 Eastern                18   
Cabbagetown, St. James Town                                      44   

                                                    Venue Category  
Neighborhood                                                        
Bathurst Quay, King and Spadina, Island airport...              16  
Berczy Park                                                     56  
Brockton, Parkdale Village, Exhibition Place                    21  
Business Reply Mail Processing Centre 969 Eastern               18  
Cabbagetown, St. James Town                                     44

In [23]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 235 uniques categories.


There are 235 uniques categories can be curated from all the returned venues. Then we analyze each neigborhood.

In [24]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       1     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Bistro  Boat or Ferry  Bookstore  \
0                0         0           0       0              0          0   
1                0         0           0       0              0          0   
2                0         0           0       0              0          0   
3                0         0           0       0              0          0   
4                0         0           0       0              0          0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0                     0               0        0                0   
1         0                     0               0        0                0   
2         0                     0               0        0                0   
3         0                     0               1        0                0   
4         0                     0               0        0                0   

   Building  Burger Joint  Burrito Place  Bus Line  Business Service  Butcher  \
0         0             0              0         0                 0        0   
1         0             0              0         0                 0        0   
2         0             0              0         0                 0        0   
3         0             0              0         0                 0        0   
4         0             0              0         0                 0        0   

   Café  Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  \
0     0                          0             0                     0   
1     0                          0             0                     0   
2     0                          0             0     

In [25]:
def column_index(df, query_cols):
    cols = df.columns.values
    sidx = np.argsort(cols)
    return sidx[np.searchsorted(cols,query_cols,sorter=sidx)]

In [26]:
column_index(toronto_onehot,['Neighborhood'])

array([159])

In [27]:
toronto_onehot = toronto_onehot.set_index(toronto_onehot.columns[159])

In [28]:
toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  \
Neighborhood                                                         
Harbourfront, Regent Park            0                  0        0   
Harbourfront, Regent Park            0                  0        0   
Harbourfront, Regent Park            0                  0        0   
Harbourfront, Regent Park            0                  0        0   
Harbourfront, Regent Park            0                  0        0   

                           Airport Food Court  Airport Gate  Airport Lounge  \
Neighborhood                                                                  
Harbourfront, Regent Park                   0             0               0   
Harbourfront, Regent Park                   0             0               0   
Harbourfront, Regent Park                   0             0               0   
Harbourfront, Regent Park                   0             0               0   
Harbourfront, Regent Park                   0             0               0   

                           Airport Service  Airport Terminal  \
Neighborhood                                                   
Harbourfront, Regent Park                0                 0   
Harbourfront, Regent Park                0                 0   
Harbourfront, Regent Park                0                 0   
Harbourfront, Regent Park                0                 0   
Harbourfront, Regent Park                0                 0   

                           American Restaurant  Antique Shop  Aquarium  \
Neighborhood                                                             
Harbourfront, Regent Park                    0             0         0   
Harbourfront, Regent Park                    0             0         0   
Harbourfront, Regent Park                    0             0         0   
Harbourfront, Regent Park                    0             0         0   
Harbourfront, Regent Park                    0             0         0   

                           Art Gallery  Arts & Crafts Store  Asian Restaurant  \
Neighborhood                                                                    
Harbourfront, Regent Park            0                    0                 0   
Harbourfront, Regent Park            0                    0                 0   
Harbourfront, Regent Park            0                    0                 0   
Harbourfront, Regent Park            0                    0                 0   
Harbourfront, Regent Park            0                    0                 0   

                           Athletics & Sports  Auto Workshop  BBQ Joint  \
Neighborhood                                                              
Harbourfront, Regent Park                   0              0          0   
Harbourfront, Regent Park                   0              0          0   
Harbourfront, Regent Park                   0              0          0   
Harbourfront, Regent Park                   0              0          0   
Harbourfront, Regent Park                   0              0          0   

                           Baby Store  Bagel Shop  Bakery  Bank  Bar  \
Neighborhood                                                           
Harbourfront, Regent Park           0           0       1     0    0   
Harbourfront, Regent Park           0           0       0     0    0   
Harbourfront, Regent Park           0           0       0     0    0   
Harbourfront, Regent Park           0           0       0     0    0   
Harbourfront, Regent Park           0           0       0     0    0   

                           Baseball Stadium  Basketball Stadium  Beach  \
Neighborhood                                                             
Harbourfront, Regent Park                 0                   0      0   
Harbourfront, Regent Park                 0                   0      0   
Harbourfront, Regent Park                 0                   0      0   
Harbourfront, Regent Park                 0  

In [29]:
toronto_onehot.shape

(1689, 234)

#### Next, group rows by neighborhood and by taking the mean of the frequency of occurrence of each category 

In [30]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0  Bathurst Quay, King and Spadina, Island airpor...     0.000000   
1                                        Berczy Park     0.000000   
2       Brockton, Parkdale Village, Exhibition Place     0.000000   
3  Business Reply Mail Processing Centre 969 Eastern     0.055556   
4                        Cabbagetown, St. James Town     0.000000   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0   0.0625              0.0625        0.0625   
1                0.0   0.0000              0.0000        0.0000   
2                0.0   0.0000              0.0000        0.0000   
3                0.0   0.0000              0.0000        0.0000   
4                0.0   0.0000              0.0000        0.0000   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0           0.125           0.1875             0.125                  0.0   
1           0.000           0.0000             0.000                  0.0   
2           0.000           0.0000             0.000                  0.0   
3           0.000           0.0000             0.000                  0.0   
4           0.000           0.0000             0.000                  0.0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0           0.0       0.0     0.000000                  0.0               0.0   
1           0.0       0.0     0.017857                  0.0               0.0   
2           0.0       0.0     0.000000                  0.0               0.0   
3           0.0       0.0     0.000000                  0.0               0.0   
4           0.0       0.0     0.000000                  0.0               0.0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                 0.0       0.000000   0.000000         0.0    0.000000   
1                 0.0       0.000000   0.017857         0.0    0.017857   
2                 0.0       0.000000   0.000000         0.0    0.000000   
3                 0.0       0.055556   0.000000         0.0    0.000000   
4                 0.0       0.000000   0.000000         0.0    0.000000   

     Bakery      Bank       Bar  Baseball Stadium  Basketball Stadium  \
0  0.000000  0.000000  0.062500               0.0            0.000000   
1  0.053571  0.000000  0.000000               0.0            0.017857   
2  0.000000  0.000000  0.047619               0.0            0.000000   
3  0.000000  0.000000  0.000000               0.0            0.000000   
4  0.045455  0.022727  0.000000               0.0            0.000000   

      Beach  Bed & Breakfast  Beer Bar  Beer Store    Bistro  Boat or Ferry  \
0  0.000000              0.0  0.000000    0.000000  0.000000         0.0625   
1  0.017857              0.0  0.035714    0.000000  0.017857         0.0000   
2  0.000000              0.0  0.000000    0.000000  0.000000         0.0000   
3  0.000000              0.0  0.000000    0.000000  0.000000         0.0000   
4  0.000000              0.0  0.000000    0.022727  0.000000         0.0000   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot   Brewery  \
0        0.0    0.0625                   0.0        0.000000  0.000000   
1        0.0    0.0000                   0.0        0.017857  0.000000   
2        0.0    0.0000                   0.0        0.095238  0.000000   
3        0.0    0.0000                   0.0        0.000000  0.055556   
4        0.0    0.0000                   0.0        0.022727  0.000000   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0              0.0       0.0           0.0       0.000000       0.0   
1              0.0       0.0           0.0       0.000000       0.0   
2              0.0       0.0           0.0       0.047619       0.0   
3              0.0       0.0           0.0       0.055556       0.0   
4              0.0       0.0           0.0       0.000000       0.0   

   Business Service   Butcher      

In [31]:
toronto_grouped.shape

(38, 235)

#### Print each neighborhood along with the top 5 most common venues

In [32]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bathurst Quay, King and Spadina, Island airport, Harbourfront West, South Niagara, Railway Lands, CN Tower----
              venue  freq
0   Airport Service  0.19
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3  Sculpture Garden  0.06
4     Boat or Ferry  0.06


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2              Bakery  0.05
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0  Breakfast Spot  0.10
1     Coffee Shop  0.10
2            Café  0.10
3             Bar  0.05
4       Pet Store  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0         Yoga Studio  0.06
1  Light Rail Station  0.06
2                Park  0.06
3          Comic Shop  0.06
4    Recording Studio  0.06


----Cabbagetown, St. James Town----
                venue  freq
0         Coffee Shop  0.07
1                 

In [33]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [34]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0  Bathurst Quay, King and Spadina, Island airpor...       Airport Service   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Parkdale Village, Exhibition Place           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4                        Cabbagetown, St. James Town           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Airport Terminal        Airport Lounge       Harbor / Marina   
1                Bakery          Cocktail Bar                  Café   
2                  Café        Breakfast Spot                   Gym   
3         Auto Workshop         Garden Center                Garden   
4                Bakery                  Park                  Café   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Coffee Shop      Sculpture Garden         Boat or Ferry   
1             Cheese Shop        Farmers Market    Seafood Restaurant   
2  Furniture / Home Store             Pet Store          Climbing Gym   
3      Light Rail Station  Fast Food Restaurant        Farmers Market   
4             Pizza Place            Restaurant                Market   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Bar              Boutique                Airport  
1              Beer Bar            Steakhouse              Irish Pub  
2            Restaurant  Caribbean Restaurant          Burrito Place  
3            Comic Shop                  Park       Recording Studio  
4                   Pub    Italian Restaurant      Indian Restaurant

#### Run *k*-means to cluster the neighborhood into 5 clusters.

In [35]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=10).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 4,
       0, 0, 0, 0, 1, 2, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0], dtype=int32)

In [36]:
neighborhoods_venues_sorted.set_index('Neighborhood').reset_index().rename({'Neighborhood': 'Neighbourhood'}, axis='columns').head()

Neighbourhood 1st Most Common Venue  \
0  Bathurst Quay, King and Spadina, Island airpor...       Airport Service   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Parkdale Village, Exhibition Place           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4                        Cabbagetown, St. James Town           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Airport Terminal        Airport Lounge       Harbor / Marina   
1                Bakery          Cocktail Bar                  Café   
2                  Café        Breakfast Spot                   Gym   
3         Auto Workshop         Garden Center                Garden   
4                Bakery                  Park                  Café   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Coffee Shop      Sculpture Garden         Boat or Ferry   
1             Cheese Shop        Farmers Market    Seafood Restaurant   
2  Furniture / Home Store             Pet Store          Climbing Gym   
3      Light Rail Station  Fast Food Restaurant        Farmers Market   
4             Pizza Place            Restaurant                Market   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Bar              Boutique                Airport  
1              Beer Bar            Steakhouse              Irish Pub  
2            Restaurant  Caribbean Restaurant          Burrito Place  
3            Comic Shop                  Park       Recording Studio  
4                   Pub    Italian Restaurant      Indian Restaurant

In [39]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_tor

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postcode           Borough              Neighbourhood   Latitude  \
2       M5A  Downtown Toronto  Harbourfront, Regent Park  43.654260   
9       M5B  Downtown Toronto   Garden District, Ryerson  43.657162   
15      M5C  Downtown Toronto             St. James Town  43.651494   
19      M4E      East Toronto                The Beaches  43.676357   
20      M5E  Downtown Toronto                Berczy Park  43.644771   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2  -79.360636               0           Coffee Shop                  Park   
9  -79.378937               0           Coffee Shop        Clothing Store   
15 -79.375418               0           Coffee Shop                 Hotel   
19 -79.293031               0     Health Food Store                 Trail   
20 -79.373306               0           Coffee Shop                Bakery   

   3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
2                    Pub                     Bakery                  Café   
9         Cosmetics Shop  Middle Eastern Restaurant                  Café   
15    Italian Restaurant                 Restaurant                  Café   
19                   Pub                      Diner        Farmers Market   
20          Cocktail Bar                       Café           Cheese Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2         Breakfast Spot    Mexican Restaurant            Restaurant   
9   Fast Food Restaurant        Lingerie Store           Pizza Place   
15        Clothing Store              Beer Bar        Breakfast Spot   
19    Falafel Restaurant           Event Space  Ethiopian Restaurant   
20        Farmers Market    Seafood Restaurant              Beer Bar   

   9th Most Common Venue       10th Most Common Venue  
2             Shoe Store            French Restaurant  
9    Japanese Restaurant                   Restaurant  
15             Gastropub                 Cocktail Bar  
19     Electronics Store  Eastern European Restaurant  
20            Steakhouse                    Irish Pub

#### Visualize resulting clustering

In [40]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Clustering 1

In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head()

Borough  Cluster Labels 1st Most Common Venue  \
2   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
19      East Toronto               0     Health Food Store   
20  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
2                   Park                   Pub                     Bakery   
9         Clothing Store        Cosmetics Shop  Middle Eastern Restaurant   
15                 Hotel    Italian Restaurant                 Restaurant   
19                 Trail                   Pub                      Diner   
20                Bakery          Cocktail Bar                       Café   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2                   Café        Breakfast Spot    Mexican Restaurant   
9                   Café  Fast Food Restaurant        Lingerie Store   
15                  Café        Clothing Store              Beer Bar   
19        Farmers Market    Falafel Restaurant           Event Space   
20           Cheese Shop        Farmers Market    Seafood Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
2             Restaurant            Shoe Store            French Restaurant  
9            Pizza Place   Japanese Restaurant                   Restaurant  
15        Breakfast Spot             Gastropub                 Cocktail Bar  
19  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant  
20              Beer Bar            Steakhouse                    Irish Pub

#### Clustering 2

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
83   Central Toronto               1            Playground   
91  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
83                  Park            Restaurant                 Diner   
91            Playground                 Trail              Building   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
83    Falafel Restaurant           Event Space  Ethiopian Restaurant   
91                 Diner    Falafel Restaurant           Event Space   

   8th Most Common Venue        9th Most Common Venue  \
83     Electronics Store  Eastern European Restaurant   
91  Ethiopian Restaurant            Electronics Store   

         10th Most Common Venue  
83          Dumpling Restaurant  
91  Eastern European Restaurant

#### Clustering 3

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto               2                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62        Ice Cream Shop           Wings Joint  Fast Food Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
62        Farmers Market    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
62  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

#### Clustering 4

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
68  Central Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
68                 Trail         Jewelry Store      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
68           Wings Joint              Dive Bar        Farmers Market   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
68    Falafel Restaurant           Event Space   Ethiopian Restaurant

#### Clustering 5

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
61  Central Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
61                  Lake              Bus Line           Swim School   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
61           Wings Joint              Dive Bar        Farmers Market   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
61    Falafel Restaurant           Event Space   Ethiopian Restaurant